src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/TextEditingTarget.java

```diff
@@ -5093,8 +5095,38 @@ protected void onChunkWritten(String chunk,
          else if (content.trim().startsWith("@"))
             wordWrap.setWrappingEnabled(true);
 
+         if (content.matches("^\\s*```")) 
+         {
+            outsideMarkdown = !outsideMarkdown;
+            wordWrap.setWrappingEnabled(outsideMarkdown);
+         }
+
+         boolean isWrappingEnabled = wordWrap.getWrappingEnabled();
+         boolean bullet = content.matches("^\\s*[-*].*");
+         if (bullet)
+            wordWrap.setWrappingEnabled(false);
+
+         boolean macroStart = content.matches("^\\s*\\\\[a-zA-Z]+\\{.*");
+         if (macroStart) 
+         {
+            macroDepth = macroDepth + 1;
+            wordWrap.setWrappingEnabled(false);
+         }
+
+         boolean macroEnd = content.matches("^\\s*\\}.*");
```

Comment:
Do we need to handle things like `\preformatted{hello}` here? Ie, things where the closing `}` isn't on its own line?

It also might be easier to choose a solution that just counts matching `{` and `}` pairs, and just keep a stack of different "states" associated with the "type" of opening `{`.